release/nightly_tests/dataset/iter_batches_benchmark.py

```diff
@@ -0,0 +1,99 @@
+import sys
+from typing import Optional
+
+import ray
+from ray.data.dataset import Dataset
+
+from benchmark import Benchmark
+
+if sys.version_info >= (3, 8):
+    from typing import Literal
+else:
+    from typing_extensions import Literal
+
+
+def iter_batches(
+    ds: Dataset,
+    batch_size: Optional[int] = None,
+    batch_format: Literal["default", "pandas", "pyarrow", "numpy"] = "default",
+    local_shuffle_buffer_size: Optional[int] = None,
+    use_default_params: bool = False,
+) -> Dataset:
+    num_batches = 0
+    if use_default_params:
+        for batch in ds.iter_batches():
+            num_batches += 1
+    else:
+        for batch in ds.iter_batches(
+            batch_format=batch_format,
+            batch_size=batch_size,
+            local_shuffle_buffer_size=local_shuffle_buffer_size,
+        ):
+            num_batches += 1
+    print("iter_batches done, num_rows:", ds.count(), "num_batches:", num_batches)
+    return ds
+
+
+def run_iter_batches_benchmark(benchmark: Benchmark):
+    ds = (
+        ray.data.read_parquet(
+            "s3://anonymous@air-example-data/ursa-labs-taxi-data/by_year/2018/01"
+        )
+        .repartition(24)
```

Comment:
One of the largest bottlenecks that we've identified in `.iter_batches()` is when batching results in a lot of slicing or concatenation, so I think that it would be good if we added some benchmarks hitting e.g. the "concatenate small blocks to create batch" case. We already have the "slice big blocks to create batch" case covered with the current setup.

Reply:
This is good tests area to cover. I think this matters for when it's distributed cluster?

Had a quick test here (single-node), the cost isn't showing up (8759874 rows, 512 blocks, so 17,108 rows/block on avg. The number in test name is the batch_size, and all of them are larger than 17,108):
```
Running case: iter-batches-block-concat-to-batch-32768
iter_batches done, num_rows: 8759874 num_batches: 268
Result of case iter-batches-block-concat-to-batch-32768: {'time': 2.7289321749994997}
Running case: iter-batches-block-concat-to-batch-65536
iter_batches done, num_rows: 8759874 num_batches: 134
Result of case iter-batches-block-concat-to-batch-65536: {'time': 2.257792303000315}
Running case: iter-batches-block-concat-to-batch-131072
iter_batches done, num_rows: 8759874 num_batches: 67
Result of case iter-batches-block-concat-to-batch-131072: {'time': 1.775239927000257}
```

Reply:
Hmm @c21 any ideas on how to better trigger the block concatenation bottleneck that y'all hit in the AIR benchmarking? I would expect it to happen when converting batches from e.g. Arrow to NumPy, and when each batch consists of 10+ blocks, maybe we're not hitting one of those conditions here that we were in the AIR benchmark.

Reply:
@jianoaix - what's the block and batch format in your quick test? For test purpose, could we try Pandas block and Pandas batch, and block:batch ratio to be ~10 (e.g. the last test case - num_batches: 67, num_blocks: 512).

Reply:
It was pyarrow block, pandas batch. Now with pandas block + pandas batch:
```
ds: Dataset(num_blocks=512, num_rows=8759874, schema={vendor_id: object, pickup_at: datetime64[ns], dropoff_at: datetime64[ns], passenger_count: int8, trip_distance: float32, rate_code_id: object, store_and_fwd_flag: object, pickup_location_id: int32, dropoff_location_id: int32, payment_type: object, fare_amount: float32, extra: float32, mta_tax: float32, tip_amount: float32, tolls_amount: float32, improvement_surcharge: float32, total_amount: float32})
Running case: iter-batches-block-concat-to-batch-32768
iter_batches done, num_rows: 8759874 num_batches: 268
Result of case iter-batches-block-concat-to-batch-32768: {'time': 6.470798211172223}
Running case: iter-batches-block-concat-to-batch-65536
iter_batches done, num_rows: 8759874 num_batches: 134
Result of case iter-batches-block-concat-to-batch-65536: {'time': 5.977799214422703}
Running case: iter-batches-block-concat-to-batch-262144
iter_batches done, num_rows: 8759874 num_batches: 34
Result of case iter-batches-block-concat-to-batch-262144: {'time': 5.614975420758128}
```